### DISCUSSION TITLE
Virtual Integration: Allow storing origin integration

### DISCUSSION THREAD

**Discussion by zweckj** [home-assistant/architecture#1160] [General] [answered]
I'm proposing we store the source virtual integration when configuring a virtual integration so it's available in the context of the serving integration's config entry.
This would give us the following possibilities:

- Avoid asking the user twice for integration information (e.g. in opower, where the user first clicks the virtual integration, but in the serving integration needs to select their provider again)
- Allow providing integration specific configuration options (e.g. for Azure OpenAI, where you need to provide additional information compared to general ChatGPT, where we don't want to confuse a "normal" user)
- Allow providing virtual integration specific branding (e.g. for Mini, which is served by BMW, but this is not obvious to the user)

During the configuration of the virtual integration, we could simply store that information in the context, so the downstream integrations can react to it.

Sample implementation PR (core): https://github.com/home-assistant/core/pull/132471

**Comment by gjohansson-ST** [comment]
While I'm supportive I think the proposal is lacking a bit on implementation details.
I think the information of the origin needs to be stored in the config entry so it's also available for configuring options, reconfiguration, reauthentication etc. and we could also show the config entry in the frontend with the right information (and even brand icons).

**Reply by zweckj** [reply]
That's what I meant with „context“. And adding it to the branding is, what I meant with the 3rd paragraph. I added „config entry“ to the context for clarification.

**Comment by frenck** [comment, marked as answer]
Hi there @zweckj 👋 

We have discussed this in an architectural core meeting; we are OK with the concept.

However, we do want to limit to to passing down the information to the integration (and maybe use it for branding, although that is also hard if one sets up 2 virtual integrations for the same supported integration).

It is kinda problematic to fully mask the origin integration, as this has much wider impact (most notably service actions).

../Frenck

**Comment by bramkragten** [comment]
Looking at the frontend implementation https://github.com/home-assistant/frontend/pull/24553, I think we miss some things for this.

The integration page of a "virtual" config entry will still be the integration page of the main integration. If you add an integration from there, you have no possibility to choose an integration/brand, it will just add an instance of the main integration without any notice of a virtual one.

How should that work?

**Reply by frenck** [reply]
I think we need to re-evaluate this decision based on this and the comments in the frontend pull request.

I've added this topic back on to the Core architectural agenda.

../Frenck
